nestkernel/event_delivery_manager.h

```diff
@@ -52,6 +52,106 @@ typedef MPIManager::OffGridSpike OffGridSpike;
 class TargetData;
 class SendBufferPosition;
 
+/*
+ * A thread-safe array to coordinate progress across threads during
+ * gather operations. Uses an array of bools instead of an STL vector,
+ * since omp atomic operations can not be applied to vector
+ * assignments.
+ */
+class CompletedChecker
+{
+private:
+  bool* a_; // must be of type bool; 'bitwise and' is used below
+            // instead of 'logical and'
+  size_t size_;
+
+public:
+  CompletedChecker()
+    : a_( NULL )
+    , size_( 0 )
+  {
+  }
+
+  ~CompletedChecker()
+  {
+    clear();
+  }
+
+  void clear()
+  {
+    VPManager::assert_single_threaded();
+    if ( a_ != NULL )
+    {
+      delete a_;
+      a_ = NULL;
+
+      assert( size_ > 0 );
+      size_ = 0;
+    }
+  }
+
+  void resize( const size_t new_size, const bool v )
+  {
+    VPManager::assert_single_threaded();
+    clear();
+
+    a_ = new bool[ new_size ];
+    for ( size_t i = 0; i < new_size; ++i )
+    {
+      a_[ i ] = v;
+    }
+    size_ = new_size;
+  }
+
+  // sets element (thread safe)
+  void set( const thread tid, const bool v )
+  {
+#pragma omp atomic write
+    a_[ tid ] = v;
+  }
+
+  // updates element (thread safe)
+  void logical_and( const thread tid, const bool v )
+  {
+#pragma omp atomic update
+    a_[ tid ] &= v;
```

Comment:
This should be `a_[tid] = a_[tid] and v;` since the operation here is a logical one, while `&=` is bitwise. Sorry I didn't catch this in my previous review.

Reply:
Technically you are right, but `and` is not supported by `omp atomic update`, see [here](https://www.ibm.com/support/knowledgecenter/SSGH2K_13.1.2/com.ibm.xlc1312.aix.doc/compiler_ref/prag_omp_atomic.html). I'll add this to the documentation.